lib/hash_ring/ets.ex

```diff
@@ -287,83 +349,184 @@ defmodule ExHashRing.HashRing.ETS do
+    Enum.reverse(found)
   end
 
-  def schedule_ring_gen_gc(state, 0), do: state
+  defp do_find_nodes(table, gen, num_nodes, remaining, hash, found, found_length) do
+    {number, node} = find_next_highest_item(table, gen, num_nodes, hash)
+
+    if node in found do
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining,
+        number,
+        found,
+        found_length
+      )
+    else
+      do_find_nodes(
+        table,
+        gen,
+        num_nodes,
+        remaining - 1,
+        number,
+        [node | found],
+        found_length + 1
+      )
+    end
+  end
+
+  defp update_nodes(%__MODULE__{} = state, nodes) do
+    new_ring_gen = state.ring_gen + 1
+
+    # Get the current generation of items from the current table
+    previous_items =
+      state.current_table
+      |> :ets.match({{state.ring_gen, :"$1"}, :"$2"})
+      |> Enum.map(fn [hash, name] ->
+        {{new_ring_gen, hash}, name}
+      end)
+
+    # Write the previous items into the previous table for this generation
+    :ets.insert(state.previous_table, previous_items)
+
+    # Generate current items for the current table
+    current_items =
+      nodes
+      |> Utils.gen_items()
+      |> Enum.map(fn {hash, name} ->
+        {{new_ring_gen, hash}, name}
+      end)
+
+    # Write the current items into the current table for this generation
+    :ets.insert(state.current_table, current_items)
 
-  def schedule_ring_gen_gc(%{pending_gcs: pending_gcs} = state, ring_gen) do
-    ring_gen_gc_delay =
-      Application.get_env(:hash_ring, :ring_gen_gc_delay, @default_ring_gen_gc_delay)
+    # Update the configuration to atomically cut over to the new generation
+    config = {
+      {state.current_table, length(nodes)},
+      {state.previous_table, length(state.nodes)},
+      new_ring_gen,
+      state.overrides
+    }
 
-    timer_ref = Process.send_after(self(), {:gc, ring_gen}, ring_gen_gc_delay)
-    %{state | pending_gcs: Map.put(pending_gcs, ring_gen, timer_ref)}
+    Config.set(state.name, self(), config)
+
+    # Schedule the previous generation for cleanup
+    state = schedule_ring_gen_gc(state, state.ring_gen)
+
+    # Update and return the state
+    %__MODULE__{state | ring_gen: new_ring_gen, nodes: nodes}
+  end
+
+  defp update_overrides(%__MODULE__{} = state, overrides) do
+    # Remove any empty overrides
+    overrides =
+      Enum.reduce(overrides, %{}, fn
+        {_, []}, acc ->
+          acc
+
+        {k, v}, acc ->
+          Map.put(acc, k, v)
+      end)
+
+    # Update the current configuration
+    case Config.get(state.name) do
+      {:ok, {current_table, previous_table, gen, _old_overrides}} ->
+        Config.set(state.name(), self(), {current_table, previous_table, gen, overrides})
+
+      _ ->
+        current_table = {state.current_table, length(state.nodes)}
+        previous_table = {state.previous_table, 0}
```

Comment:
This is to handle the case where config isn't aware of the current config before we set an override? That seems invalid, and we should error?

Reply:
yea, I can just crash the process in this case

Reply:
Crash behavior implemented in https://github.com/discord/ex_hash_ring/pull/10/commits/173fbda6fff6a11075af73ad466eed548991667b